File: python/leetcode/160. Intersection of Two Linked Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def getIntersectionNode(self, headA: ListNode,
                            headB: ListNode) -> ListNode:
        if headA is None or headB is None:
            return None

        first = headA
        second = headB

        while first != second:
            first = first.next
            second = second.next

            if first == second:
                return first

            if not first:
                first = headB
            if not second:
                second = headA
        return first

File: c/leetcode/160. Intersection of Two Linked Lists.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *getIntersectionNode(struct ListNode *headA, struct ListNode *headB)
{
  if (!headA || !headB)
    return NULL;

  struct ListNode *first = headA;
  struct ListNode *second = headB;

  while (first != second)
  {
    first = first->next;
    second = second->next;

    if (first == second)
      return first;

    if (!first)
      first = headB;
    if (!second)
      second = headA;
  }

  return first;
}